### DISCUSSION TITLE
HowTo: Complete Guide to manualy install text-generation-webui + Vicuna 1.1 4bit 13B 128g (or any other 4bit LLM) localy with Windows WSL & Ubuntu for 8GB or higher GPU

### DISCUSSION THREAD

**Discussion by eucdee** [oobabooga/text-generation-webui#1797] [General]
### (or any other 4bit LLM) localy with Windows 10 WSL & Ubuntu 22.04

the current Linux WSL guide is out of date, so here is a working version from 05/05/2023
please note that things develop exponentialy quicker with a.i. 💯 

### 1. About:
[**Vicuna 1.1**](https://vicuna.lmsys.org/) is a state-of-art Large Language Model (LLM)
our version has the weights alreadly applied and is quantized to 4bit to lower the GPU RAM requirements. thanks @TheBloke 

im using Windows 10 but this should also work on Windows 11
we will download pytorch 2.0.0 with cuda 11.7 support, it should also work with nightly and cuda 11.8, or later versions.
custom versions can be found [here](https://pytorch.org/get-started/locally/)

we are using latest [GPTQ-for-LLaMa triton branch](https://github.com/qwopqwop200/GPTQ-for-LLaMa#readme) to make it possible.
[here](https://github.com/oobabooga/text-generation-webui/blob/main/docs/GPTQ-models-(4-bit-mode).md) are some more infos about GPTQ-for-LLaMa

i should note im using a 3080 notebook gpu with 8GB
inference is not much faster than the local windows 1click install i first used from [troublechute](https://hub.tcno.co/ai/text-ai/vicuna/) 
the fastest version is still via kaggle or free google colab, you can use my [**notebook, its also configured for vicuna 1.1**](https://colab.research.google.com/github/eucdee/AI/blob/main/4bit_TextGen_Gdrive.ipynb)

you will need the nvidia cuda toolkit for windows, [latest version](https://developer.nvidia.com/cuda-toolkit-archive) should be good. 
to make sure i have everything on the same version i used [CUDA Toolkit 11.8](https://developer.nvidia.com/cuda-11-8-0-download-archive?target_os=Windows&target_arch=x86_64&target_version=10&target_type=exe_local) but as you will find out, ive used a older pytorch version. maybe inference is faster on later versions..

### 2. Windows Prequisities:
**Download & Install** [latest CUDA Toolkit for Windows ](https://developer.nvidia.com/cuda-toolkit-archive)

### 3. Setup WSL:
Basicly you open powershell and put in the following commands:
```
wsl --install
wsl --set-default-version 2
```
then go to the Windows Store, search, download and open Ubuntu 22.04 (or latest version, maybe even different distribution) 

please note that you need at least Windows 10.
[here](https://github.com/oobabooga/text-generation-webui/blob/main/docs/WSL-installation-guide.md) is a in-depth guide in case you dont know how to do this.

After opening you will need to choose username & password for your linux wsl.

### 4. Ubuntu Terminal Commands:
now just follow these commands and it should work, 
```
sudo apt update && sudo apt -y upgrade
curl -sL "https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh" > "Miniconda3.sh"
bash Miniconda3.sh
```
Now close the window, the terminal needs to restart.
while the terminal is down go and edit the **.bashrc** with windows explorer and editor:
`\\wsl.localhost\Ubuntu-22.04\home\$USER\`

copy and paste the following to the bottom:
```
export CUDA_HOME=/home/$USER/miniconda3/envs/textgen/pkgs/cuda-toolkit/
export PATH="$PATH:/home/$USER/miniconda3/envs/textgen/bin/"
export LD_LIBRARY_PATH="/usr/lib/wsl/lib:/home/$USER/miniconda3/lib/"
```

now open ubuntu terminal and continue with the command line:
```
sudo apt install build-essential
conda create -n textgen python=3.10.9
conda activate textgen
conda install pytorch torchvision torchaudio pytorch-cuda=11.7 -c pytorch -c nvidia
conda install -c conda-forge cudatoolkit-dev
git clone https://github.com/oobabooga/text-generation-webui
cd text-generation-webui
pip install -r requirements.txt
mkdir repositories
cd repositories
git clone https://github.com/qwopqwop200/GPTQ-for-LLaMa.git -b triton
cd GPTQ-for-LLaMa
pip install ninja
pip install -r requirements.txt
cd
cd text-generation-webui
```

### 5. Download Model:
now to download a model, in our case Vicuna 1.1 we use this command:
`python download-model.py TheBloke/vicuna-13B-1.1-GPTQ-4bit-128g`

### 6. Launch:
to launch everything we need this command, you should customize the arguments, find more info [here](https://github.com/oobabooga/text-generation-webui)
`python server.py --auto-devices --chat --threads 12 --wbits 4 --groupsize 128 --pre_layer 20 --gpu-memory 7`

this is it, visit `http://127.0.0.1:7860/` and start talking...

after reboot you will always need to activate conda textgen first:
```
conda activate textgen
cd text-generation-webui
python server.py --auto-devices --chat --threads 12 --wbits 4 --groupsize 128 --pre_layer 20 --gpu-memory 7
```

now let me try and get [deepspeed](https://github.com/oobabooga/text-generation-webui/blob/main/docs/DeepSpeed.md) working to add to the guide.

### 7. Misc:
here are some helpfull commands to debug the install:

```
nvcc -version
nvidia-smi
python -m torch.utils.collect_env
```

also this python script [cuda_check.py](https://gist.github.com/f0k/63a664160d016a491b2cbea15913d549)

much thanks to @cmhamiche for helping me get this running and everyone else involved in this mindblowing piece of software.

**Comment by Catherine-Delacroix** [comment]
Did you manage to get deepspeed working? Never managed to run it with quantized models

**Reply by eucdee** [reply]
> Did you manage to get deepspeed working? Never managed to run it with quantized models

i think it doesnt work with gptq or ggml models yet

**Comment by mironkraft** [comment]
I would like to ask you, if in windows, you have the enought knowledge to help or guide to setup the next thing based on this:

1. Developping a interface on UI so you can drag and drop documents
2. Creating a folder where to store as documents, information from where the AI will learn or acquire information 

Or in linux, it does not matter, both options are cool

**Comment by codeninjasvaughan** [comment]
When I try to run python server.py --auto-devices --chat --threads 12 --wbits 4 --groupsize 128 --pre_layer 20 --gpu-memory 7, the process eventually ends with 'Killed'. Any ideas?

**Reply by EdSeymore** [reply]
My experience: that means insufficient memory.

**Comment by miles-du** [comment]
Thank you very much for taking the time to share your work! I am having some issues with my install and may have a shot with your guide soon! Meanwhile:

 - "this is it, visit http://127.0.0.1:7860/ and start talking..."

Hi, I'd love to know how this step works for you? Since installing Ooba in WSL this cannot work from Windows I have found. The Linux machine uses a virtual NIC that does not respond to the Windows machine localhost. I have tried the suggested:
"netsh interface portproxy add v4tov4 listenaddress=0.0.0.0 listenport=7860 connectaddress=localhost connectport=7860"
but no luck.
So far it only works using the --share argument, which is not satisfactory.

Any help appreciated!

**Reply by mironkraft** [reply]
Won't work AFAIK, you should download then windows version

**Reply by miles-du** [reply]
I have been using the Windows version for 2 months. I will find a solution to this issue, it's  not time to just give up!

**Comment by Woisek** [comment]
Great stuff! Any chance to get TheBloke/vicuna-13B-1.1-GPTQ-4bit-128g loaded on a 8GB card? 🤔
Or is there an optimized version available?
